From: =?utf-8?q?Michel_D=C3=A4nzer?= <mdaenzer@redhat.com>
Date: Thu, 6 Mar 2025 17:20:38 +0100
Subject: kms/impl-device: Don't set_needs_flush if there's a pending page
 flip

Just make sure notify_crtc_frame_ready will call
meta_kms_impl_device_schedule_process again when the flip completes.

This is sufficient to avoid
https://gitlab.gnome.org/GNOME/mutter/-/issues/3002 .

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/3002
Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/3955
Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/3959
Bug-Debian: https://bugs.debian.org/1099743
Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4321
---
 src/backends/native/meta-kms-impl-device.c | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/src/backends/native/meta-kms-impl-device.c b/src/backends/native/meta-kms-impl-device.c
index 73ac6cf..3b32ba4 100644
--- a/src/backends/native/meta-kms-impl-device.c
+++ b/src/backends/native/meta-kms-impl-device.c
@@ -67,6 +67,7 @@ typedef struct _CrtcDeadline
   MetaKmsCrtc *crtc;
   MetaKmsUpdate *pending_update;
   gboolean await_flush;
+  gboolean need_schedule;
   gboolean pending_page_flip;
 
   struct {
@@ -1423,7 +1424,7 @@ notify_crtc_frame_ready (CrtcFrame *crtc_frame)
   crtc_frame->pending_page_flip = FALSE;
   crtc_frame->deadline.is_deadline_page_flip = FALSE;
 
-  if (!crtc_frame->pending_update)
+  if (!crtc_frame->pending_update && !crtc_frame->need_schedule)
     return;
 
   if (crtc_frame->await_flush)
@@ -2070,17 +2071,19 @@ meta_kms_impl_device_schedule_process (MetaKmsImplDevice *impl_device,
 
   crtc_frame = ensure_crtc_frame (impl_device, crtc);
 
+  if (crtc_frame->pending_page_flip)
+    {
+      crtc_frame->need_schedule = TRUE;
+      return;
+    }
+  crtc_frame->need_schedule = FALSE;
+
   if (crtc_frame->await_flush)
     return;
 
-  if (is_using_deadline_timer (impl_device))
-    {
-      if (crtc_frame->pending_page_flip)
-        return;
-
-      if (ensure_deadline_timer_armed (impl_device, crtc_frame))
-        return;
-    }
+  if (is_using_deadline_timer (impl_device) &&
+      ensure_deadline_timer_armed (impl_device, crtc_frame))
+    return;
 
   meta_kms_device_set_needs_flush (meta_kms_crtc_get_device (crtc), crtc);
 }
